aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -424,6 +497,14 @@ impl AptosVM {
 
         charge_global_write_gas_usage(gas_status, &session, &txn_data.sender())?;
 
+        // call init function of the each module
+        self.execute_module_initialization(
+            &mut session,
+            gas_status,
+            modules,
+            &[txn_data.sender()],
+        )?;
```

Comment:
this is probably not what you expect, returning Error means internal invariant violation. a failed transaction should return Ok(VMStatus::Error(_), output) and need to deal with gas charge cleanup as well

Reply:
double checked this. This is intended. the reasons are 

1. I think the gas is still charged when performing `failed_transaction_cleanup_and_keep_vm_status`. 
2. `failed_transaction_cleanup_and_keep_vm_status` will create a new session which will not save the module if the init_module failed, which is desired. 
3. keep it consistent with other function in execute_modules, eg: `Self::verify_module_bundle`